cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.ql

```diff
@@ -12,29 +12,69 @@
 
 import cpp
 import semmle.code.cpp.commons.Exclusions
-import semmle.code.cpp.valuenumbering.GlobalValueNumbering
 import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
+import semmle.code.cpp.rangeanalysis.RangeAnalysisUtils
 import semmle.code.cpp.controlflow.Guards
+import semmle.code.cpp.dataflow.DataFlow
 
-/** Holds if `sub` is guarded by a condition which ensures that `left >= right`. */
+/**
+ *  Holds if `sub` is guarded by a condition which ensures that
+ * `left >= right`.
+ */
 pragma[noinline]
 predicate isGuarded(SubExpr sub, Expr left, Expr right) {
-  exists(GuardCondition guard |
-    guard.controls(sub.getBasicBlock(), true) and
-    guard.ensuresLt(left, right, 0, sub.getBasicBlock(), false)
+  exists(GuardCondition guard, int k |
+    guard.controls(sub.getBasicBlock(), _) and
+    guard.ensuresLt(left, right, k, sub.getBasicBlock(), false) and
+    k >= 0
   )
 }
 
-/** Holds if `sub` will never be negative. */
-predicate nonNegative(SubExpr sub) {
-  not exprMightOverflowNegatively(sub.getFullyConverted())
+/**
+ * Holds if `e` is known or suspected to be less than or equal to
+ * `sub.getLeftOperand()`.
+ */
+predicate exprIsSubLeftOrLess(SubExpr sub, Expr e) {
+  e = sub.getLeftOperand()
+  or
+  exists(Expr other |
+    // use-use
+    exprIsSubLeftOrLess(sub, other) and
+    (
+      useUsePair(_, other, e) or
+      useUsePair(_, e, other)
```

Comment:
I don't think `useUsePair` is used anywhere else in the LGTM or Code Scanning suites. At least I couldn't find it in `runSnapshotQueries-debug.log` of our latest Analyze-Snapshots job. The `useUsePair` predicate is inherently quadratic, so there's a risk we'll time out on any project that's unlucky enough to have many uses without intervening definitions.

Reply:
I wasn't aware of that - I'll check later how much we lose if we have only the `localFlowStep` case.

Reply:
Pushed a fix, see comment below.  I think it's a better and cleaner solution, so thanks for pointing this out!

Reply:
Thanks for the fix. I haven't considered any other aspects of this PR, so I'll hand it back to @MathiasVP for reviewing.